src/main/scala/co/teapot/tempest/algorithm/MonteCarloPPR.scala

```diff
@@ -23,52 +23,117 @@ import scala.concurrent.{Await, Future}
+  *     walk of unconstrained length.
+  *  On a user-item bipartite graphs, if you want to find similar items to a given item, see the tweet producer case above.
+  *  On a user-item bipartite graphs, if you want to find similar users to a given user, see the tweet consumer case above.
+  *  On a user-item bipartite graphs, if you want to recommend items to a given user, see the who-to-follow case above.
+  *
+  *  Background reading: "WTF: the who to follow service at Twitter" by Gupta, Goel, Lin, Sharma, Wnag, Zadeh.
   */
 object MonteCarloPPR {
-  def estimatePPR(graph: DirectedGraph, startIds: IndexedSeq[Int], params: MonteCarloPageRankParams,
+  def estimatePPR(graph: DirectedGraph,
+                  startIds: IndexedSeq[Int],
+                  firstStepDirection: EdgeDir,
+                  lengthConstraint: LengthConstraint,
+                  params: MonteCarloPageRankParams,
                   random: Random = new Random()):
   collection.Map[Int, Double] = {
     val pprs = CollectionUtil.efficientIntDoubleMapWithDefault0()
     def randomStart(): Int = startIds(random.nextInt(startIds.size))
+
+    var currentDirection = firstStepDirection
+    var currentLength = 0
     var v = randomStart()
+
     for (stepIndex <- 0 until params.getNumSteps) {
-      pprs(v) += 1.0 / params.getNumSteps
-      if (random.nextDouble() < params.getResetProbability) {
+      // Make sure our walk length is acceptable (e.g. even or odd according to lengthConstraint).  If not, don't record
+      // that we visited this node.
+      if (lengthConstraint.isAllowed(currentLength)) {
+        pprs(v) += 1.0 / params.getNumSteps
+      }
+
+      val vDegree = graph.degree(v, currentDirection)
+      if (random.nextDouble() < params.getResetProbability || vDegree == 0) {
         v = randomStart()
+        currentDirection = firstStepDirection
+        currentLength = 0
       } else {
-        val vDegree = graph.outDegree(v)
-        if (vDegree > 0)
-          v = graph.outNeighbor(v, random.nextInt(vDegree))
-        else
-          v = randomStart()
+        v = graph.neighbor(v, random.nextInt(vDegree), currentDirection)
+        if (params.alternatingWalk) {
+          currentDirection = currentDirection.flip
+        }
+        currentLength += 1
       }
     }
-    if (params.isSetMinReportedVisits) {
-      pprs retain {
+    filterPPRResults(pprs, params)
+  }
+
+  /** See estimatePPR for an explanation of the arguments. */
+  def estimatePPRParallel(graph: DirectedGraph,
+                          startIds: IndexedSeq[Int],
+                          firstStepDirection: EdgeDir,
+                          lengthConstraint: LengthConstraint,
+                          params: MonteCarloPageRankParams,
+                          threadCount: Int):
+  collection.Map[Int, Double] = {
+    val paramsForThread = params.deepCopy()
+    paramsForThread.setNumSteps(params.getNumSteps / threadCount) // Each thread will do it's share of the steps.
```

Comment:
Good catch.  That's always been a pet peeve of mine: if Pranav owns a thing, it is Pranav's, so if it owns a thing, clearly it should be it's.  Unfortunately that's not a battle I'll win anytime soon.